### PR TITLE
CS1 videos UX

### PR DESCRIPTION
**Pull request by shubhi1092**
- Added videos link on world map and PlayLevelView for students playing CS1
- Videos link opens a new videos modal with locked and unlocked status for each video
- Changed course victory modal text and behavior when next view is the video view
- Changed styling of buttons on victory modal
- Added new PlayLevelVideoView to show the video

### CODE REVIEW

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+</template>
+
+<script>
+import utils from 'core/utils'
+import FlatLayout from 'core/components/FlatLayout'
+import Player from '@vimeo/player'
+
+export default Vue.extend({
+  props: {
+    levelID: {  // level-slug
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoComponent.vue]: Should we rename this `levelSlug` then?

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      required: true
+    },
+    courseID: {
+      type: String,
+      required: true
+    },
+    codeLanguage: {
+      type: String
+    },
+    level: {    // level original id
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoComponent.vue]: Should we rename this `levelOriginalId`?

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  data: () => ({
+    videoData: {},
+    startTime: '',
+    endTime: ''
+  }),
+  components: {
+    'flat-layout': FlatLayout
+  },
+  created() {
+    this.videoData = utils.videoLevels[this.level]
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoComponent.vue]: Do we need to handle a mismatch here?

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            :href="nextLevelLink"
+          )
+            | {{ $t('play_level.next_level') }}
+      img.img-unlocked(src = '/images/pages/play/modal/unlocked_banner.png')
+</template>
+
+<script>
+import utils from 'core/utils'
+import FlatLayout from 'core/components/FlatLayout'
+import Player from '@vimeo/player'
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoComponent.vue]: Should we call this `VideoPlayer` instead, since we have player users as well?

File: app/views/play/level/PlayLevelVideoView.coffee [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+utils = require 'core/utils'
+
+module.exports = class PlayLevelVideoView extends RootComponent
+  id: 'play-level-video-view'
+  template: require 'templates/base-flat'
+  VueComponent: PlayLevelVideoComponent
+
+  constructor: (options, @levelID) ->
+    @propsData = { @levelID }
+    @propsData.courseID = utils.getQueryVariable 'course' or null
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoView.coffee]: `getQueryVariable` 2nd param is default value:
https://github.com/codecombat/codecombat/blob/b77b425/app/core/utils.coffee#L376

Also, why explicit null instead of undefined here?

File: app/views/play/level/modal/CourseVictoryComponent.vue [hunk 6]
```diff
@@ -28,7 +28,7 @@
               div {{ $t('play_level.combo_challenge_complete_body', { concept: primaryConcept }) }}
                   
             div.clearfix.well.well-sm.well-parchment(v-else-if="assessmentNext")
-              img.lock-banner(src="/images/pages/play/modal/lock_banner.png")
+              img.lock-banner(src="/images/pages/play/modal/unlocked_banner.png")
```
Comment (differentmatt) [inline app/views/play/level/modal/CourseVictoryComponent.vue]: Why this change?

File: app/views/play/level/modal/CourseVictoryComponent.vue [hunk 7]
```diff
@@ -129,11 +141,16 @@
             link = "/play/level/#{@nextLevel.slug}?course=#{@course._id}&course-instance=#{@courseInstanceID}"
             link += "&codeLanguage=" + @level.primerLanguage if @level.primerLanguage
           return link
+      nextLevelVideoLink: ->
+        link = "/play/video/level/#{@nextLevel.slug}?course=#{@course._id}&course-instance=#{@courseInstanceID}"
+        link += "&codeLanguage=" + @level.primerLanguage if @level.primerLanguage
+        link += "&level=" + @nextLevel.original
+        return link
       nextLevelLinkClasses: ->
```
Comment (differentmatt) [inline app/views/play/level/modal/CourseVictoryComponent.vue]: Seems kinda weird to put class selection in computed rather than the template directly.  No need to change this now.

File: app/views/play/level/modal/CourseVideosModalComponent.vue [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    courseID: {
+      type: String,
+      required: true
+    }
+  },
+  data: () => ({
+    videoLevels: [],
+    vimeoPlayer: {}
+  }),
+  created() {
```
Comment (differentmatt) [inline app/views/play/level/modal/CourseVideosModalComponent.vue]: Can we use `co` to unwind these nested promises into cleaner yield calls?
https://github.com/codecombat/codecombat/blob/b77b425/app/views/admin/AdminClassroomLevelsComponent.vue#L63

File: app/views/play/level/modal/CourseVideosModalComponent.vue [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+          for (let l of this.videoLevels){
+            if (sessionsMap[l.original]){
+              Vue.set(l, 'videoStatus', 'unlocked')
+            }
+            else{
+              Vue.set(l, 'videoStatus', 'locked')
+            }
+          }
+        }).catch((err) => {
+          console.log("Error in fetching sessions:", err)
```
Comment (differentmatt) [inline app/views/play/level/modal/CourseVideosModalComponent.vue]: Should these be console.error?

File: app/views/play/level/modal/CourseVideosModalComponent.vue [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+      console.log("Error in fetching course instance details:", err)
+    });
+  },
+  methods: {
+    // open video in the iframe when video thumbnail is clicked
+    onImageClick: function(e) {
+      let video_url = ''
+      let src = e.target.src
+      src = src.slice(src.search('/images'))
+      for (let l of this.videoLevels){
```
Comment (differentmatt) [inline app/views/play/level/modal/CourseVideosModalComponent.vue]: `let video_url = (this.videoLevels.find(l => l.thumbnail_unlocked == src) || {}).url`?

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        $('#next-level-btn')[0].style.display = "block"
+      })
+      this.startTime = new Date()
+      // hack to remove base template's header and footer
+      $('#main-nav')[0].style.display = "none"
+      $('#footer')[0].style.display = "none"
+      $('#final-footer')[0].style.display = "none"
+    })
+  },
+  beforeDestroy () {
```
[the remark this answers is not in the document]
Comment (shubhi1092) [reply app/views/play/level/PlayLevelVideoComponent.vue]: Yep this is a vue lifecycle method that is called everytime the vue component gets destroyed, i.e. everytime we navigate away from this page. (https://vuejs.org/v2/api/#beforeDestroy)

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  },
+  mounted() {
+    this.$nextTick(function () {
+      const player = new Player($('.video')[0]);
+      player.on('ended', function() {
+        $('#next-level-btn')[0].style.display = "block"
+      })
+      this.startTime = new Date()
+      // hack to remove base template's header and footer
+      $('#main-nav')[0].style.display = "none"
```
[the remark this answers is not in the document]
Comment (shubhi1092) [reply app/views/play/level/PlayLevelVideoComponent.vue]: Yeah, makes sense. Will store their values to make them visible again.

File: app/views/play/level/PlayLevelVideoView.coffee (hunk 5 above)
Comment (shubhi1092) [reply app/views/play/level/PlayLevelVideoView.coffee]: Yes, even with `null`, it is being set as `undefined` because of the second parameter of `getQueryVariable` which is fine, so will remove explicit `null`.

File: app/views/play/level/modal/CourseVictoryComponent.vue (hunk 6 above)
Comment (shubhi1092) [reply app/views/play/level/modal/CourseVictoryComponent.vue]: I renamed this image to be `unlocked_banner` since it is for unlocked state, and there is another image for locked status which is `locked_banner`. So I renamed it to make it clear.

File: app/views/play/level/modal/CourseVideosModalComponent.vue (hunk 8 above)
Comment (shubhi1092) [reply app/views/play/level/modal/CourseVideosModalComponent.vue]: For ES6 transpilation, we do not support yields and generators yet. We might need to install another babel library to support and build that code. Will change it if its fairly simple.

File: app/views/play/level/PlayLevelVideoComponent.vue [hunk 13]
```diff
@@ -77,20 +78,20 @@ export default Vue.extend({
...
-        'footer': $('#footer')[0].style.display,
-        'final-footer': $('#final-footer')[0].style.display
+        'main-nav': $('#main-nav')[0]? $('#main-nav')[0].style.display : "",
+        'footer': $('#footer')[0]? $('#footer')[0].style.display : "",
+        'final-footer': $('#final-footer')[0]? $('#final-footer')[0].style.display: ""
       }
-      $('#main-nav')[0].style.display = "none"
-      $('#footer')[0].style.display = "none"
-      $('#final-footer')[0].style.display = "none"
+      if ($('#main-nav')[0]) { $('#main-nav')[0].style.display = "none" }
```
Comment (differentmatt) [inline app/views/play/level/PlayLevelVideoComponent.vue]: If we're going to do this on one line, can remove `{ }` function body.

File: test/app/views/play/level/PlayLevelVideoView.spec.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+          return locale.en.translation[text]; 
+        }
+      }
+    }
+  })
+}
+
+const props = {
+  // a level id with the video details
+  levelOriginalID: utils.videoLevels ? Object.keys(utils.videoLevels)[0] || '54173c90844506ae0195a0b4' : '54173c90844506ae0195a0b4',
```
Comment (differentmatt) [inline test/app/views/play/level/PlayLevelVideoView.spec.js]: `levelOriginalID: Object.keys(utils.videoLevels || {})[0] || '54173c90844506ae0195a0b4',`

File: test/app/views/play/level/modal/CourseVideosModalComponent.spec.js [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import videosModalComponent from 'views/play/level/modal/CourseVideosModalComponent'
+import locale from 'locale/locale'
+import CourseInstance from 'collections/CourseInstances'
+import Course from 'collections/Courses'
+import Levels from 'collections/Levels'
+import factories from 'test/app/factories'
+import utils from 'core/utils'
+import api from 'core/api'
+import LevelSession from 'collections/LevelSessions'
+// import $ from 'jquery'
```
Comment (differentmatt) [inline test/app/views/play/level/modal/CourseVideosModalComponent.spec.js]: Can remove?

File: test/app/views/play/level/modal/CourseVideosModalComponent.spec.js [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+  })
+}
+
+// create levels that have the video details
+const lev = []
+for (let l in utils.videoLevels){
+  lev.push(factories.makeLevel({original: l}))
+}
```
Comment (differentmatt) [inline test/app/views/play/level/modal/CourseVideosModalComponent.spec.js]: `const lev = utils.videoLevels.map((l) => factories.makeLevel({original: l}))`

File: test/app/views/play/level/modal/CourseVideosModalComponent.spec.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        else {
+          return locale.en.translation[text]; }
+        }
+    }
+  })
+}
+
+// create levels that have the video details
+const lev = []
```
Comment (differentmatt) [inline test/app/views/play/level/modal/CourseVideosModalComponent.spec.js]: Rename to `levelTestData`?

**Comment by shubhi1092** [comment]
Thank you for the code review! :)
